File: app/utils/s3_uploader.py
import os
import boto3
from botocore.exceptions import NoCredentialsError, ClientError

class S3Uploader:
    def __init__(self):
        # 자격 증명 로드
        self.access_key = os.getenv("AWS_ACCESS_KEY_ID")
        self.secret_key = os.getenv("AWS_SECRET_ACCESS_KEY")
        self.region = os.getenv("AWS_DEFAULT_REGION")
        
        # 타겟 ARN (버킷 대신 사용)
        self.target_arn = os.getenv("AWS_S3_ACCESS_POINT_ARN")
        
        # [수정] 경로 프리픽스 로드 (없으면 빈 문자열)
        self.base_prefix = os.getenv("S3_BASE_PREFIX", "")
        self.app_prefix = os.getenv("S3_APP_PREFIX", "")

        # 클라이언트 초기화
        self.s3_client = boto3.client(
            's3',
            aws_access_key_id=self.access_key,
            aws_secret_access_key=self.secret_key,
            region_name=self.region
        )

    def upload_file(self, file_content: bytes, file_name: str, folder: str = "regulation") -> str:
        """
        [경로 조립 로직 추가]
        Base Prefix + App Prefix + Category Folder + Filename
        """
        if not self.target_arn:
            print("❌ AWS_S3_ACCESS_POINT_ARN 설정이 없습니다.")
            return None
            
        # 1. 경로 조립 (빈 값은 걸러내고 /로 연결)
        # 예: team17/raw_data/regulation/hash.pdf
        path_parts = [self.base_prefix, self.app_prefix, folder, file_name]
        # 빈 문자열('')이나 None을 제거하고 '/'로 합침
        s3_key = "/".join([p for p in path_parts if p])
        
        try:
            print(f"☁️ S3 업로드 시도: {s3_key}")
            
            # 2. 업로드 수행
            self.s3_client.put_object(
                Bucket=self.target_arn,
                Key=s3_key,
                Body=file_content,
                ContentType='application/pdf' if file_name.endswith('.pdf') else 'text/plain'
            )
            
            # 3. 저장된 전체 경로 반환 (DB 저장용)
            # Access Point URL이 있다면 앞에 붙여줄 수도 있지만, 
            # 내부 식별용으로는 Key만 있어도 충분하거나 s3:// 프로토콜 사용
            return f"s3://{s3_key}"
            
        except NoCredentialsError:
            print("❌ AWS 자격 증명 오류")
            return None
        except ClientError as e:
            print(f"❌ S3 업로드 실패 (ClientError): {e}")
            return None
        except Exception as e:
            print(f"❌ 알 수 없는 업로드 오류: {e}")
            return None

File: application_service/src/main/java/com/global17/auth_service/util/S3Uploader.java
package com.global17.auth_service.util;

import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;

@Component
public class S3Uploader {

    @Value("${aws.accessKeyId}")
    private String accessKey;
    @Value("${aws.secretAccessKey}")
    private String secretKey;
    @Value("${aws.region}")
    private String region;
    
    @Value("${aws.s3.target-arn}")
    private String targetArn;
    
    @Value("${aws.s3.base-prefix}")
    private String basePrefix;
    @Value("${aws.s3.app-prefix}")
    private String appPrefix;

    private S3Client s3Client;

    @PostConstruct
    public void init() {
        this.s3Client = S3Client.builder()
                .region(Region.of(region))
                .credentialsProvider(StaticCredentialsProvider.create(
                        AwsBasicCredentials.create(accessKey, secretKey)))
                .build();
    }

    public String uploadFile(byte[] content, String fileName, String category) {
        // 경로 조립: base/app/category/filename
        String s3Key = String.format("%s/%s/%s/%s", basePrefix, appPrefix, category, fileName)
                .replaceAll("//+", "/"); // 중복 슬래시 제거

        try {
            System.out.println("☁️ S3 Uploading: " + s3Key);
            
            s3Client.putObject(PutObjectRequest.builder()
                    .bucket(targetArn) // Access Point ARN 사용
                    .key(s3Key)
                    .build(), RequestBody.fromBytes(content));
            
            return "s3://" + s3Key;
        } catch (Exception e) {
            System.err.println("❌ S3 Upload Failed: " + e.getMessage());
            return null;
        }
    }
}




